sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala

```diff
@@ -300,6 +300,67 @@ class EventTimeWatermarkSuite extends StreamTest with BeforeAndAfter with Matche
     )
   }
 
+  test("watermark with 2 streams") {
+    val first = MemoryStream[Int]
+
+    val firstAggregation = first.toDF()
+      .withColumn("eventTime", $"value".cast("timestamp"))
+      .withWatermark("eventTime", "10 seconds")
+      .select('value)
+
+    val second = MemoryStream[Int]
+
+    val secondAggregation = second.toDF()
+      .withColumn("eventTime", $"value".cast("timestamp"))
+      .withWatermark("eventTime", "10 seconds")
+      .select('value)
+
+    val union = firstAggregation.union(secondAggregation)
+      .writeStream
+      .format("memory")
+      .queryName("test")
+      .start()
+
+    def populateNewWatermarkFromData(stream: MemoryStream[Int], data: Int*): Unit = {
+      stream.addData(data)
+      union.processAllAvailable()
+      // add a dummy batch so lastExecution has the new watermark
+      stream.addData(0)
+      union.processAllAvailable()
+    }
+
+    def assertQueryWatermark(watermark: Int): Unit = {
+      assert(union.asInstanceOf[StreamingQueryWrapper].streamingQuery
+        .lastExecution.offsetSeqMetadata.batchWatermarkMs
+        == watermark)
+    }
+
+    populateNewWatermarkFromData(first, 11)
+    assertQueryWatermark(1000)
+
+    // Watermark stays at 1 from the left when right watermark moves to 2
```

Comment:
nit: Watermark to "global watermark" .. to differentiate from "right watermark" later in the sentence.